raft/src/test/java/org/apache/kafka/raft/MockLog.java

```diff
@@ -220,46 +230,68 @@ public LogAppendInfo appendAsFollower(Records records) {
     }
 
     public List<LogBatch> readBatches(long startOffset, OptionalLong maxOffsetOpt) {
-        long maxOffset = maxOffsetOpt.orElse(endOffset().offset);
-        if (startOffset > maxOffset) {
-            throw new OffsetOutOfRangeException("Requested offset " + startOffset + " is larger than " +
-                "the provided end offset " + maxOffsetOpt);
-        }
+        verifyOffsetInRange(startOffset);
 
+        long maxOffset = maxOffsetOpt.orElse(endOffset().offset);
         if (startOffset == maxOffset) {
             return Collections.emptyList();
         }
 
         return log.stream()
-            .filter(batch -> batch.firstOffset() >= startOffset && batch.lastOffset() < maxOffset)
+            .filter(batch -> batch.lastOffset() >= startOffset && batch.lastOffset() < maxOffset)
```

Comment:
Is this intentional?

Reply:
Yeah, I think this was a bug. We need to include the batch if it contains `startOffset`. The previous check only worked when the fetch offset is aligned at the start of the batch. This is consistent with the logic in `read`.

Reply:
Acked the `Log` and confirmed, I was originally thinking that if a fetch start offset is in the middle of a batch (in practice it should not happen at least from our own Java clients though) we would start from the next batch. Realized it's not the case actually :)